### PR TITLE
Removed MauticFactory::getDatabase.

### PR DESCRIPTION
**Pull request by biozshock**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 🔴 
| New feature/enhancement? (use the a.x branch)      | 🟢
| Deprecations?                          | 🟢
| BC breaks? (use the c.x branch)        | 🟢
| Automated tests included?              | 🟢 

## Description
Continue to shave off MauticFactory. The getDatabase is removed.

---
### 📋 Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Try to install and/or update a plugin.

### CODE REVIEW

File: app/bundles/PluginBundle/Bundle/PluginDatabase.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+use Doctrine\ORM\Mapping\ClassMetadata;
+use Doctrine\ORM\Tools\SchemaTool;
+use Mautic\IntegrationsBundle\Migration\Engine;
+use Mautic\PluginBundle\Entity\Plugin;
+
+class PluginDatabase
+{
+    private readonly string $mauticDbPrefix;
+
+    public function __construct(private readonly EntityManagerInterface $em, private readonly Connection $connection, ?string $mauticDbPrefix)
```
Comment (escopecz) [inline app/bundles/PluginBundle/Bundle/PluginDatabase.php]: Do you think we could try this? https://symfony.com/blog/new-in-symfony-6-3-dependency-injection-improvements#new-options-for-autowire-attribute
```suggestion
    public function __construct(
         private readonly EntityManagerInterface $em,
         private readonly Connection $connection,
         #[Autowire(env: 'DB_TABLE_PREFIX')]
         private readonly string $mauticDbPrefix
     )
```
It's more readable than adding it to services.php

Comment (biozshock) [reply app/bundles/PluginBundle/Bundle/PluginDatabase.php]: @escopecz Thanks. Applied the suggestion with proper env variable.

File: app/bundles/PluginBundle/Bundle/PluginDatabase.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        #[Autowire(env: 'MAUTIC_TABLE_PREFIX')]
+        ?string $mauticDbPrefix,
+    ) {
+        $this->mauticDbPrefix = $mauticDbPrefix ?? '';
+    }
+
+    /**
+     * Install plugin schema based on Doctrine metadata.
+     *
+     * @psalm-param array<ClassMetadata> $metadata
```
Comment (escopecz) [inline app/bundles/PluginBundle/Bundle/PluginDatabase.php]: I'm wondering why psalm?
```suggestion
     * @param array<ClassMetadata> $metadata
```

**Review by escopecz — COMMENTED** [review]
I will have to find a plugin to test this with.

File: app/bundles/PluginBundle/Bundle/PluginDatabase.php (hunk 2 above)
Comment (biozshock) [reply app/bundles/PluginBundle/Bundle/PluginDatabase.php]: A copy from the base class. Didn't want to interfere there to support the phpdoc later.

Comment (escopecz) [reply app/bundles/PluginBundle/Bundle/PluginDatabase.php]: Apologies, I don't get the reason. Can you rephrase?

Comment (biozshock) [reply app/bundles/PluginBundle/Bundle/PluginDatabase.php]: NVM, mixed up with another change :)

**Comment by biozshock** [comment]
@escopecz Created a plugin for this matter: https://github.com/biozshock/mautic-db-plugin-test. I'm not sure how i can put this one into the automatic testing process yet. But at least the removal of the database is testable now.

**Comment by escopecz** [comment]
@biozshock thanks for the plugin! I guess we could also use the custom-objects plugin for the same.

I was able to install the plugin and the table showed up:
Screenshot 2025-02-05 at 10 05 47

But when I switched the plugin tag to 1.1.0 and run the command again it failed:
```
bin/console m:p:u
10:06:04 ERROR     [mautic] Doctrine\DBAL\Driver\PDO\PDOException: There is no active transaction (uncaught exception) at /var/www/html/vendor/doctrine/dbal/src/Driver/PDO/PDOException.php line 20 while running console command `mautic:plugins:reload`
[stack trace]
#0 /var/www/html/vendor/doctrine/dbal/src/Driver/PDO/Connection.php(128): Doctrine\DBAL\Driver\PDO\PDOException::new()
#1 /var/www/html/vendor/doctrine/dbal/src/Driver/Middleware/AbstractConnectionMiddleware.php(78): Doctrine\DBAL\Driver\PDO\Connection->commit()
#2 /var/www/html/vendor/doctrine/dbal/src/Logging/Connection.php(70): Doctrine\DBAL\Driver\Middleware\AbstractConnectionMiddleware->commit()
#3 /var/www/html/vendor/doctrine/dbal/src/Driver/Middleware/AbstractConnectionMiddleware.php(78): Doctrine\DBAL\Logging\Connection->commit()
#4 /var/www/html/vendor/symfony/doctrine-bridge/Middleware/Debug/DBAL3/Connection.php(109): Doctrine\DBAL\Driver\Middleware\AbstractConnectionMiddleware->commit()
#5 /var/www/html/vendor/doctrine/dbal/src/Connection.php(1464): Symfony\Bridge\Doctrine\Middleware\Debug\DBAL3\Connection->commit()
#6 /var/www/html/vendor/doctrine/dbal/src/Connection.php(1422): Doctrine\DBAL\Connection->doCommit()
#7 /var/www/html/vendor/doctrine/orm/src/EntityManager.php(291): Doctrine\DBAL\Connection->commit()
#8 /var/www/html/app/bundles/IntegrationsBundle/Migration/Engine.php(50): Doctrine\ORM\EntityManager->commit()
#9 /var/www/html/app/bundles/PluginBundle/Bundle/PluginDatabase.php(80): Mautic\IntegrationsBundle\Migration\Engine->up()
#10 /var/www/html/app/bundles/PluginBundle/EventListener/PluginSubscriber.php(35): Mautic\PluginBundle\Bundle\PluginDatabase->onPluginUpdate()
#11 /var/www/html/vendor/symfony/event-dispatcher/Debug/WrappedListener.php(116): Mautic\PluginBundle\EventListener\PluginSubscriber->onUpdate()
#12 /var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php(220): Symfony\Component\EventDispatcher\Debug\WrappedListener->__invoke()
#13 /var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php(56): Symfony\Component\EventDispatcher\EventDispatcher->callListeners()
#14 /var/www/html/vendor/symfony/event-dispatcher/Debug/TraceableEventDispatcher.php(139): Symfony\Component\EventDispatcher\EventDispatcher->dispatch()
#15 /var/www/html/app/bundles/PluginBundle/Helper/ReloadHelper.php(85): Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher->dispatch()
#16 /var/www/html/app/bundles/PluginBundle/Facade/ReloadFacade.php(33): Mautic\PluginBundle\Helper\ReloadHelper->updatePlugins()
#17 /var/www/html/app/bundles/PluginBundle/Command/ReloadCommand.php(34): Mautic\PluginBundle\Facade\ReloadFacade->reloadPlugins()
#18 /var/www/html/vendor/symfony/console/Command/Command.php(326): Mautic\PluginBundle\Command\ReloadCommand->execute()
#19 /var/www/html/vendor/symfony/console/Application.php(1096): Symfony\Component\Console\Command\Command->run()
#20 /var/www/html/vendor/symfony/framework-bundle/Console/Application.php(126): Symfony\Component\Console\Application->doRunCommand()
#21 /var/www/html/vendor/symfony/console/Application.php(324): Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand()
#22 /var/www/html/vendor/symfony/framework-bundle/Console/Application.php(80): Symfony\Component\Console\Application->doRun()
#23 /var/www/html/vendor/symfony/console/Application.php(175): Symfony\Bundle\FrameworkBundle\Console\Application->doRun()
#24 /var/www/html/bin/console(16): Symfony\Component\Console\Application->run()
#25 {main} ["hostname" => "mautic-community-web","pid" => 2298]
10:06:04 CRITICAL  [console] Error thrown while running command "'m:p:u'". Message: "There is no active transaction" ["exception" => Doctrine\DBAL\Driver\PDO\PDOException^ { …},"command" => "'m:p:u'","message" => "There is no active transaction"] ["hostname" => "mautic-community-web","pid" => 2298]
10:06:04 WARNING   [mautic] Command `mautic:plugins:reload` exited with status code 1 ["hostname" => "mautic-community-web","pid" => 2298]

In PDOException.php line 20:
                                  
  There is no active transaction  
                                  

In Connection.php line 126:
                                  
  There is no active transaction  
                                  

mautic:plugins:reload
```

**Comment by biozshock** [comment]
@escopecz Well, then it was not working before :)
The issue is the same as here: https://github.com/doctrine/DoctrineMigrationsBundle/issues/393
Solution commit it seems: https://github.com/doctrine/migrations/commit/c4c46f7064f6e7795bd7f26549579918b46790fa

**Comment by biozshock** [comment]
@escopecz The issue was fixed in the 5.2. Should be OK to merge now.

**Review by escopecz — APPROVED** [review]
Yes! No error triggered now. The table got the new "field" column when upgraded to the version 1.1.0 🎉 

Thank you for helping me test this!
